polkadot/runtime/parachains/src/inclusion/mod.rs

```diff
@@ -879,15 +879,10 @@ impl<T: Config> Pallet<T> {
 		// initial weight is config read.
 		let mut weight = T::DbWeight::get().reads_writes(1, 0);
 		if let Some(new_code) = commitments.new_validation_code {
-			// Block number of candidate's inclusion.
-			let now = <frame_system::Pallet<T>>::block_number();
-
-			weight.saturating_add(<paras::Pallet<T>>::schedule_code_upgrade(
+			weight.saturating_accrue(Self::try_schedule_code_upgrade(
```

Comment:
I just realized that this is too late. We need to drop the candidate in backing already, but in a way that should be safe. Mainly there exists a mechanism for this already: [UpgradeRestrictionSignal](https://github.com/paritytech/polkadot-sdk/blob/69434d9a32af6103d7cbde9798cf594027fb620e/polkadot/runtime/parachains/src/paras/mod.rs#L800). If an upgrade is not allowed for whatever reason this signal should be set. This is read by the parachain and it won't even try (which is mandatory otherwise it would be bricked, because we would then filter the candidate [here](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L1306), called [here](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L1250) from [verify_backed_candidate](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L1194).

Reason being: If we only drop the upgrade in inclusion we did not actually prevent the spam vector, as most of the work has already been done. Most importantly a relay chain block was already filled with megabytes of data, affecting other candidates.

Also we should update the guide as well. In particular I forgot again why we need fees at all:

1. Deposit is large for large runtimes.
2. Frequency of updates is limited by cool down.
3. Block validation is already being paid for.

I think the major reason is that upgrading a runtime is rather costly, not only in terms of block utilization (size), but also because of pre-checking. This essentially breaks sharding (all validators do it). So it makes sense to be a bit cautious here. The more important part of this PR is surely that we adjust the deposit.

Reply:
In order to not delay this further, let's make this a follow up.

Reply:
> I just realized that this is too late. We need to drop the candidate in backing already, but in a way that should be safe. Mainly there exists a mechanism for this already: [UpgradeRestrictionSignal](https://github.com/paritytech/polkadot-sdk/blob/69434d9a32af6103d7cbde9798cf594027fb620e/polkadot/runtime/parachains/src/paras/mod.rs#L800). If an upgrade is not allowed for whatever reason this signal should be set. This is read by the parachain and it won't even try (which is mandatory otherwise it would be bricked, if we only had the filtering of candidates [here](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L1306), called [here](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L1250) from [verify_backed_candidate](https://github.com/paritytech/polkadot-sdk/blob/e1c033ebe15e6de5ee6a5a81eab2dd213f09d574/polkadot/runtime/parachains/src/inclusion/mod.rs#L1194)).

This is not really true, `UpgradeRestrictionSignal` is for telling the parachain that it is currently not allowed to do an upgrade. However, we currently have no way to signal that we want to do an upgrade to the relay chain without sending the full new wasm. You are right that the spam vector isn't closed with the current implementation and we should fix it. I mean we could maybe abuse `UpgradeRestrictionSignal`, but then we should entirely drop the dynamic reserve based on the size of the wasm blob. We should instead charge a constant fee for this (probably just based on the maximum blob size). Then we could use the `UpgradeRestrictionSignal` to communicate if the billing account has the required `UpgradeFee`. But yeah, still not a super solution. Someone could still announce an upgrade and move the funds in between backing and availability.

The real proper solution is really to rewrite the current signaling to the relay chain to first only announce the size plus hash of the blob. Then we should reserve the amount on the billing account and signal that it is okay to send the code.

Reply:
Fixed deposit would be indeed a lot simpler, I see two downsides with it:

1. People are not incentivized to save space.
2. If we ever increase the limit those earlier parachains can now increase their size for free.

both are not dramatic and also if we would ever find those to be a problem we can still write code to adjust it on upgrades.

I hate to say this, but maybe we should just adjust the deposit to fixed maximum for now and get back to this later, e.g. with me actually writing a solid design for this first (proper signaling, actual concerns, potential solutions, actually picked solution with reason) and ideally also then also clean up the code while we are at it.

Reply:
Probably a dumb question, but what if we restricted upgrades to only happen from the registrar pallet through the extrinsic? This way the spam vector wouldn't be possible.

Reply:
Can't a parachain make an XCM call to the relay chain to schedule a code upgrade?

Reply:
How would that make a difference? Even if we bumped XCM message size limitations to account for that, we would have the exact same situation, just wrapped differently.

Reply:
Yeah, I had an idea, but realized it wouldn't work.

What if we put the parachain whose upgrade schedule failed into cooldown? @eskimor as you mentioned the block validation is already being paid for, so from the perspective of the rc it shouldn't really matter if the upgrade failed or succeeded, and with the cooldown we prevent spamming.
Also, we can probably remove the `UpgradeFee`.

Reply:
Reminds me of my idea [here](https://github.com/paritytech/polkadot-sdk/issues/669#issuecomment-1763325926) :P

Reply:
But yeah, if we do this. Then we don't need to change any of the signaling logic etc.

Reply:
Great, wanted to get this confirmed :). I will implement this tomorrow then.

Reply:
For sure we should not use any kind of dynamic fee based on wasm size for the deposit. This fee should be based on the maximum size of the wasm file.

Reply:
We have the ‘UpgradeFee’ and the deposit. I am not sure I understand what problems would the dynamic deposit cause if we put the para into cooldown regardless whether the scheduling succeeded or failed.

Reply:
I mean the logic in this pr would not be required anymore? 

But we should require the max at registration, the point being that people otherwise register a very small blob and then upgrade to something much bigger.

Reply:
I was thinking of adding this to this PR. It would actually be a minimal change(inside the inclusion module put the para into cooldown if calling  `try_schedule_code_upgrade` failed).

This way we would have an overall robust system with:
- dynamic deposits to incentivize people to save space. This also overall makes more sense since you only pay for the space you use.
- cooldown to prevent spamming
- `UpgradeFee` is implemented to cover the cost of pre-checking. One might argue that this is not necessary since we have a cooldown, but I believe it makes sense for the following reason: The cooldown only limits the frequency at which upgrades can be scheduled. Without the `UpgradeFee`, it would cost the same to schedule code upgrades at the interval of the cooldown duration as it would to schedule upgrades less frequently. With the `UpgradeFee`, a parachain that schedules upgrades more frequently will incur higher costs.

Reply:
> dynamic deposits to incentivize people to save space. This also overall makes more sense since you only pay for the space you use.

We need to tackle this entire topic any way in a different way. As you remember Gav and Shawn weren't that happy about your state rent model for the validation code. So, we will need to come up with something better there, which will change whatever we would do here any way. I don't see a reason to make this more complicated right now.

> UpgradeFee to cover the cost of pre-checking.

The entire reason why this was is too prevent spamming. The spamming would be prevented by the timeout. So, I don't see any reason to add a second "barrier". On the other side this makes the "user" (aka parachains) experience worse. They will need to have  some billing account and whatever. If possible, I would prevent adding fees for upgrades. We would also need to update tutorials etc. This is quite a lot of work for no real benefit IMO. 

Maybe let's see what @eskimor thinks.

Reply:
Yeah, I don't think I presented the state rent model well on the call. The focus seemed to be on the removal of the validation code for parachains that did not pay the rent. The main argument against it was that it would 'break' parachains. However, if we consider that to be breaking couldn't we say the same thing for the Coretime model? If a parachain doesn't purchase Coretime it can't produce blocks. I think as long as the `RentDuration` is set to a sensible value like 4 weeks there shouldn't be problems.

Shawn described an alternative solution that would involve the addition of new social constructs rather than introducing a new model. While this could be effective, I believe an automated process might be preferable to avoid the extra social overhead for a problem that could be handled by code automatically.

Overall I think it would be good if the rent model RFC could get some more reviews.

Reply:
> Maybe let's see what @eskimor thinks.

I had similar thoughts: Is the complexity really worth it? 

@szegoo has a very good point about incentivization: If the upgrade costs something, then we incentivize teams to not "needlessly" upgrade, which is very sensible as there is a significant cost for the network. But I like @bkchr idea of having a very long cooldown period (which is enforced also on failed upgrades) and then you could optionally pay to reduce the cooldown. This would get us all the benefits, without risking irritation about having to set up some account and stuff. People would only need to look into it once they are unhappy with the cooldown time.

So the simplest way forward would be:

# Phase 1:

1. Make the deposit always account for the maximum code size.
2. Increase cooldown period .... actually this is also something we would need to communicate ... but likely less irritating than suddenly having to setup an account.
3. Make cooldown to also be enforced on failed updates.

# Phase 2:

Then we can have a second (not time critical) phase:

This phase is then merely an optimization, as opposed to a necessity. Hence we should also write an RFC here (design document), highlighting exactly all the requirements and options considered.


4. Find good solution for deposit size: Having it dependent on actual size is a good thing in general as it just aligns incentives with cost, leading to better utilization of resources.
5. Implement/adjust what we already have in this PR to enable being able to pay for a reduced cooldown.
6. See downsides: For rent based model/reduced deposit fees, upgrade fees might be mandatory.

# Downsides

If we don't charge for upgrades (usually), then this also increases pressure on deposits not getting too low. As deposits then serve two purposes:

1. Limit state used (obviously).
2. Limit number of upgrades that can happen (together with cooldown)

Phase 1: Should be fast to implement ... if not, maybe we can rely on the large deposit for now and move (2) and (3)  to phase 2 as well. Just changing the deposit calculation (making it bigger) should not even require a security audit.

Reply:
For the rent based model. Indeed there was quite some confusing on pruning, I don't think Gav nor Shawn understood what was actually meant here. Pruning was specifically designed to not be able to break anything - at least not anymore than an ending lease would for example.

The real question whether the rent based model is feasible is, whether we are for security mostly relying on the opportunity costs of the provided deposit or the actually locked capital. In a very first analysis of mine, it is unfortunately in part also the amount of locked capital: Basically if you fill up the storage, managing to degrade the service enough, this would have an impact on the price of DOT, so your deposited DOT would decrease in value -> Disincentivizing the attack.

Reply:
For moving forward in launching coretime on Kusama, @Szegoo would you be interested into implementing  phase 1? Phase 1 MVP, would entail only changing the deposit amount to the maximum PVF size.

Reply:
> 3\. Make cooldown to also be enforced on failed updates.

This is already the case, at least I checked this last week. [We set the cooldown before we kick off the pvf check](https://github.com/paritytech/polkadot-sdk/blob/c83ccda4a32c8b54a57accb1a874337ec1e368cf/polkadot/runtime/parachains/src/paras/mod.rs#L1931-L1943).

Reply:
>  @Szegoo would you be interested into implementing phase 1? 

Yes, I will open a PR for that.